synapse/storage/databases/main/roommember.py

```diff
@@ -751,7 +751,6 @@ def do_users_share_a_room_txn(
                     SELECT room_id, state_key FROM current_state_events
                     WHERE type = 'm.room.member' AND membership = 'join' AND {clause}
                 ) AS b using (room_id)
-                LIMIT 1
```

Comment:
Can you explain the cause of the bug and why this fixes it?

Reply:
If you get 2 people with presence updates that have disjoint rooms in common with the user syncing, this method will only return one of them.

Removing the limit fixes that, and the surrounding code is already dealing with having multiple values in the response.

Reply:
I think the `LIMIT 1` should be in the subselect which defines `b`? (We don't care about how many rooms Alice and Bob share---we just want to know if they share _some_ room.)

Reply:
Suppose Alice and Bob share say 40 rooms. Without `LIMIT 1`, looking up users who share rooms with Alice is going to send back Bob 40 times over.

For example:

```
matrix=> SELECT b.state_key
matrix-> FROM (
matrix(>     SELECT room_id FROM current_state_events
matrix(>     WHERE type = 'm.room.member' AND membership = 'join' AND state_key = '@dmrobertson:matrix.org'
matrix(> ) AS a
matrix-> INNER JOIN (
matrix(>     SELECT room_id, state_key FROM current_state_events
matrix(>     WHERE type = 'm.room.member' AND membership = 'join' AND state_key = ANY('{@erikj:jki.re,@doesnotexist:example.com}'::text[])
matrix(> ) AS b using (room_id)
matrix-> ;
   state_key   
═══════════════
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
 @erikj:jki.re
(69 rows)

Time: 5.173 ms

```

Reply:
I spent some time playing around with this. The following works for postgres, but not sqlite as it uses a lateral join.

```
matrix=> WITH other_users(other_user_id) AS (VALUES ('@erikj:jki.re'), ('@doesnotexist:example.com'), ('@richvdh:sw1v.org'))
matrix-> SELECT other_user_id, t.*
matrix-> FROM other_users
matrix->     JOIN LATERAL (
matrix(>         SELECT EXISTS (
matrix(>             SELECT 1
matrix(>             FROM current_state_events a
matrix(>                 JOIN current_state_events b USING (room_id)
matrix(>             WHERE a.type = 'm.room.member' AND a.membership = 'join' AND a.state_key = '@dmrobertson:matrix.org'
matrix(>               AND b.type = 'm.room.member' AND b.membership = 'join' AND b.state_key = other_user_id
matrix(>         )
matrix(>     ) t ON true
matrix-> ;
       other_user_id       │ exists 
═══════════════════════════╪════════
 @erikj:jki.re             │ t
 @doesnotexist:example.com │ f
 @richvdh:sw1v.org         │ t
(3 rows)

Time: 2.903 ms
```

Reply:
Just adding a `DISTINCT` to the first select seems to work ? It's not completely optimized but at least we don't get the penalty of parsing and iterating on them in Python.

Reply:
Doesn't like there's much in it. But DISTINCT seems slightly faster (see below), is clearer and (presumably works on sqlite too). Distinct SGTM!

<details>
<summary>For completeness</summary>

```
davidr@mooncalf:~/bench$ cat distinct.sql 
SELECT DISTINCT b.state_key
FROM (
    SELECT room_id FROM matrix.current_state_events
    WHERE type = 'm.room.member' AND membership = 'join' AND state_key = '@dmrobertson:matrix.org'
) AS a
INNER JOIN (
    SELECT room_id, state_key FROM matrix.current_state_events
    WHERE type = 'm.room.member' AND membership = 'join' AND state_key IN (SELECT name FROM matrix.users LIMIT 1000)
) AS b using (room_id);
davidr@mooncalf:~/bench$ pgbench -f distinct.sql --transactions=100 matrix
starting vacuum...ERROR:  relation "pgbench_branches" does not exist
(ignoring this error and continuing anyway)
ERROR:  relation "pgbench_tellers" does not exist
(ignoring this error and continuing anyway)
ERROR:  relation "pgbench_history" does not exist
(ignoring this error and continuing anyway)
end.
transaction type: distinct.sql
scaling factor: 1
query mode: simple
number of clients: 1
number of threads: 1
number of transactions per client: 100
number of transactions actually processed: 100/100
latency average = 197.492 ms
tps = 5.063491 (including connections establishing)
tps = 5.064597 (excluding connections establishing)



davidr@mooncalf:~/bench$ cat lateral.sql 
WITH other_users(other_user_id) AS (SELECT name FROM matrix.users LIMIT 1000)
SELECT other_user_id, t.*
FROM other_users
    JOIN LATERAL (
        SELECT EXISTS (
            SELECT 1
            FROM matrix.current_state_events a
                JOIN matrix.current_state_events b USING (room_id)
            WHERE a.type = 'm.room.member' AND a.membership = 'join' AND a.state_key = '@dmrobertson:matrix.org'
              AND b.type = 'm.room.member' AND b.membership = 'join' AND b.state_key = other_user_id
        )
    ) t ON true
;
davidr@mooncalf:~/bench$ pgbench -f lateral.sql --transactions=100 matrix
starting vacuum...ERROR:  relation "pgbench_branches" does not exist
(ignoring this error and continuing anyway)
ERROR:  relation "pgbench_tellers" does not exist
(ignoring this error and continuing anyway)
ERROR:  relation "pgbench_history" does not exist
(ignoring this error and continuing anyway)
end.
transaction type: lateral.sql
scaling factor: 1
query mode: simple
number of clients: 1
number of threads: 1
number of transactions per client: 100
number of transactions actually processed: 100/100
latency average = 210.455 ms
tps = 4.751608 (including connections establishing)
tps = 4.752533 (excluding connections establishing)
```
</details>